dbms/src/DataStreams/HashJoinProbeBlockInputStream.cpp

```diff
@@ -41,34 +47,54 @@ Block HashJoinProbeBlockInputStream::getHeader() const
     assert(res.rows() == 0);
     ProbeProcessInfo header_probe_process_info(0);
     header_probe_process_info.resetBlock(std::move(res));
-    return join->joinBlock(header_probe_process_info);
-}
-
-void HashJoinProbeBlockInputStream::finishOneProbe()
-{
-    bool expect = false;
-    if likely (probe_finished.compare_exchange_strong(expect, true))
-        join->finishOneProbe();
+    /// use original_join here so we don't need add lock
+    return original_join->joinBlock(header_probe_process_info);
 }
 
 void HashJoinProbeBlockInputStream::cancel(bool kill)
 {
     IProfilingBlockInputStream::cancel(kill);
     /// When the probe stream quits probe by cancelling instead of normal finish, the Join operator might still produce meaningless blocks
     /// and expects these meaningless blocks won't be used to produce meaningful result.
-    try
+
+    JoinPtr current_join;
+    BlockInputStreamPtr current_non_joined_stream;
+    BlockInputStreamPtr current_restore_build_stream;
+    BlockInputStreamPtr current_restore_probe_stream;
     {
-        finishOneProbe();
+        std::lock_guard lock(mutex);
+        current_join = join;
+        current_non_joined_stream = non_joined_stream;
+        current_restore_build_stream = restore_build_stream;
+        current_restore_probe_stream = restore_probe_stream;
     }
-    catch (...)
+    /// Cancel join just wake up all the threads waiting in Join::waitUntilAllBuildFinished/Join::waitUntilAllProbeFinished,
+    /// the ongoing join process will not be interrupted
+    /// There is a little bit hack here because cancel will be called in two cases:
+    /// 1. the query is cancelled by the caller or meet error: in this case, wake up all waiting threads is safe
+    /// 2. the query is executed normally, and one of the data stream has read an empty block, the the data stream and all its children
+    ///    will call `cancel(false)`, in this case, there is two sub-cases
+    ///    a. the data stream read an empty block because of EOF, then it means there must be no threads waiting in Join, so cancel the join is safe
+    ///    b. the data stream read an empty block because of early exit of some executor(like limit), in this case, just wake the waiting
+    ///       threads is not 100% safe because if the probe thread is wake up when build is not finished yet, it may produce wrong results, for now
+    ///       it is safe because when any of the data stream read empty block because of early exit, the execution framework ensures that no further
+    ///       data will be used.
+    current_join->cancel();
```

Comment:
How about just calling `current_join->cancel();` here?
And check `isCancelled()` in `while (restore_build_stream->read()) {};` and etc.

Reply:
You mean don't cancel the related stream? But if `HashJoinProbeBlockInputStream::cancel()` is called, there is no guarantee that `HashJoinProbeBlockInputStream::readImpl()` will be called again.